_posts/2025-02-28-lattice-bitset.md

```diff
@@ -0,0 +1,801 @@
+<path fill="none" stroke="black" d="M40.01,-108.3C40.01,-115.59 40.01,-124.27 40.01,-132.46"/>
+<polygon fill="black" stroke="black" points="36.51,-132.38 40.01,-142.38 43.51,-132.38 36.51,-132.38"/>
+</g>
+</g>
+</svg>
+</figure>
+
+Where *N* represents some integer stored in the `value` field.
+
+This complicates things a bit. Let's put both the type bits and the
+specialization together in one structure and admire:
+
+```c++
+struct Type {
+    enum TypeBits {
+        Bottom = 0,       // 0b000
+        Int    = 1 << 0,  // 0b001
+        List   = 1 << 1,  // 0b010
+        String = 1 << 2,  // 0b100
+        Top    = 7,       // 0b111
+    } type;
+    // If you are feeling clever, you can also steal some `type` bits for the
+    // `spec_kind` instead of putting it in a separate field.
+    struct Spec {
+        enum {
+            SpecTop,
+            SpecInt,
+            SpecBottom,
+        } spec_kind;
+        int value;
+    } spec;
+};
+
+struct Type TTop = (struct Type) {
+    .type = Top,
+    .spec = (struct Spec) { .spec_kind = SpecTop },
+};
+struct Type TInt = (struct Type) {
+    .type = Int,
+    .spec = (struct Spec) { .spec_kind = SpecTop },
+};
+// Invariant: .type == Bottom if and only if .spec_kind == SpecBottom.
+struct Type TBottom = (struct Type) {
+    .type = Bottom,
+    .spec = (struct Spec) { .spec_kind = SpecBottom },
+};
+```
+
+That's very nice and more precise, but now our `join` and `is_subtype`
+operators don't make a whole lot of sense. We can't just use bitwise operations
+any more. We have to also do the lattice operations on the `Spec` field:
+
+```c++
+struct Type join(struct Type left, struct Type right) {
+    struct Type result;
+    result.type = left.type | right.type;
+    result.spec = spec_join(left.spec, right.spec);
+    return result;
+}
+
+// Ask if `left` is a subtype of `right`
+bool is_subtype(struct Type left, struct Type right) {
+    return (left & right) == left && spec_is_subtype(left.spec, right.spec);
+}
+```
+
+If we decompose the problem that way, we can write some lattice operations for
+`Spec`. Let's start with `is_subtype`:
+
+```c++
+// Ask if `left` is a subtype of `right`
+bool spec_is_subtype(struct Spec left, struct Spec right) {
+    if (right.spec_kind == SpecTop || left.spec_kind == SpecBottom) {
+        // Top is a supertype of everything and Bottom is a subtype of
+        // everything
+        return true;
+    }
+    // left is now either SpecInt or SpecTop
+    // right is now either SpecInt or SpecBottom
+    // The only way left could be a subtype of right is if they are both
+    // SpecInt and the value is the same
+    if (left.spec_kind == SpecInt &&
+        right.spec_kind == SpecInt &&
+        left.value == right.value) {
+        return true;
+    }
+    return false;
+}
+```
+
+That takes a bit of time to internalize, so please read over it a couple of
+times and work out the cases by hand. It's really useful for implementing
+`spec_join`:
```

Comment:
I didn't personally find this part very hard to internalize. Not particularly suggesting you change this paragraph, just letting you know that it's clear and easy to follow.